### PR TITLE
Tweak OpenBSD cpu detection

### PR DESCRIPTION
**Pull request by jasperla**
In OpenBSD 6.4 a new sysctl was added to enable/disable SMT on the
fly (hw.smt). At the same time a new sysctl was added which reflects
the number of online CPUs/cores (hw.ncpuonline).

num_cpus was using the hw.ncpu sysctl value which includes the number
of offline cores when hw.smt=0. Adjust num_cpus to return the actual
number of usable cores with get() and implement get_num_physical_cpus()
which returns the total number of CPUs/cores in the system.

### CODE REVIEW

**Comment by seanmonstar** [comment]
Cool! One question: does this change make it so num_cpus would start failing on OpenBSD if the version is before 6.4?

**Comment by semarie** [comment]
OpenBSD 6.4 is unsupported on OpenBSD side, and so, previous versions are unsupported too.

If someone would run recent num_cpus on such version, `sysctl` call will failed. I suppose (not checked) that `cpus` will stay to `0`, and so the function will returns `1`.

**Comment by semarie** [comment]
from [sysctl(2)](https://man.openbsd.org/sysctl.2) man page, it seems that `oldp` could be altered and `sysctl(2)` return `-1` (and set `errno`).

For me, it would be better to check `sysctl(2)` returned value to fallback to `cpus=1` instead of the fact that `cpus < 1`.

@jasperla could you update your PR with that ?

**Comment by jasperla** [comment]
@seanmonstar sorry for the delay -- I have updated this PR to incorporate @semarie's suggestion. I have also applied it to the existing `sysctl` call for OpenBSD as that could potentially fail (for whichever reason) so the code is concistent for this platform.

**Comment by seanmonstar** [comment]
> OpenBSD 6.4 is unsupported on OpenBSD side, and so, previous versions are unsupported too.

I don't understand what you mean. `num_cpus::get()` had OpenBSD support, checking `libc::HW_NCPU`. This changes it to check `libc::HW_NCPUONLINE`. My question is: will this make `num_cpus` fail to compile on OpenBSD <6.4? Or return 1 when it used to return `NCPU`?

**Comment by jasperla** [comment]
> I don't understand what you mean. `num_cpus::get()` had OpenBSD support, checking `libc::HW_NCPU`. This changes it to check `libc::HW_NCPUONLINE`. My question is: will this make `num_cpus` fail to compile on OpenBSD <6.4? Or return 1 when it used to return `NCPU`?

On OpenBSD 6.4, before my changes, `num_cpus::get()` was returning the total number of CPUs/cores present in the system through `hw.ncpu`. This was incorrect behaviour. And `get_num_physical_cpus()` was not implemented for OpenBSD yet.

After my changes, on OpenBSD 6.4, `num_cpus::get()` will return `1` as the `hw.ncpuonline` call will fail.  `num_cpus::get_num_physical_cpus()` on OpenBSD 6.4 will now use `hw.ncpu` and thus return what `num_cpus::get()` was returning before my changes.

On OpenBSD 6.5 and later (i.e. all supported OpenBSD releases) have the correct behaviour for `num_cpus::get()` and `num_cpus::get_num_physical_cpus()`.

Does this clear the confusion?

**Comment by seanmonstar** [comment]
Yep, thank you! Sounds great!

File: src/lib.rs [hunk 1]
```diff
@@ -307,19 +307,43 @@ fn get_num_cpus() -> usize {
...
-                     0);
+        rc = libc::sysctl(mib.as_mut_ptr(),
+                          2,
+                          &mut cpus as *mut _ as *mut _,
+                          &mut cpus_size as *mut _ as *mut _,
+                          ptr::null_mut(),
+                          0);
     }
-    if cpus < 1 {
+    if rc < 0 {
```
Comment (seanmonstar) [inline src/lib.rs]: Is there never a case where `sysctl` could succeed, but leave the number at 0?

Comment (jasperla) [reply src/lib.rs]: Not according to the [manpage](https://man.openbsd.org/sysctl.2#RETURN_VALUES) which explicitly states:

> If the call to sysctl() is unsuccessful, -1 is returned and errno is set appropriately.

**Comment by jasperla** [comment]
Thank you @semarie  for your input and @seanmonstar for the merge!

**Pull request merged**